skorch/tests/test_doctor.py

```diff
@@ -0,0 +1,781 @@
+
+        num_activation_recs_after = len(doctor.activation_recs_['module'])
+        num_gradient_recs_after = len(doctor.gradient_recs_['module'])
+
+        assert num_activation_recs_before == num_activation_recs_after
+        assert num_gradient_recs_before == num_gradient_recs_after
+
+    def test_callbacks_cleaned_up_after_fit_with_initial_callbacks(
+            self, doctor_cls, net_cls, module_cls, data
+    ):
+        # make sure that the callbacks are the same before and after, this is
+        # important because SkorchDoctor will temporarily add a callback
+        from skorch.callbacks import EpochScoring, GradientNormClipping
+
+        net = net_cls(
+            module_cls,
+            callbacks=[EpochScoring('f1'), GradientNormClipping(1.0)],
+        ).initialize()
+        callbacks_without_doctor = net.callbacks_[:]
+
+        doctor = doctor_cls(net).fit(*data)
+        callbacks_with_doctor = doctor.net.callbacks_
+
+        assert len(callbacks_without_doctor) == len(callbacks_with_doctor)
+
+        for (name0, cb0), (name1, cb1) in zip(
+                callbacks_without_doctor, callbacks_with_doctor
+        ):
+            assert name0 == name1
+            # pylint: disable=unidiomatic-typecheck
+            assert type(cb0) == type(cb1)
+
+    def test_get_layer_names(self, doctor):
+        layer_names = doctor.get_layer_names()
+        expected = {
+            'criterion': [],
+            'module': ['lin0', 'lin1', 'softmax']
+        }
+        assert layer_names == expected
+
+    def test_get_parameter_names(self, doctor):
+        param_names = doctor.get_param_names()
+        expected = {
+            'criterion': [],
+            'module': ['lin0.weight', 'lin0.bias', 'lin1.weight', 'lin1.bias'],
+        }
+        assert param_names == expected
+
+    def test_predict(self, doctor, data):
+        X, _ = data
+        y_pred_doctor = doctor.predict(X)
+        y_pred_net = doctor.net.predict(X)
+        np.testing.assert_allclose(y_pred_doctor, y_pred_net)
+
+    def test_predict_proba(self, doctor, data):
+        X, _ = data
+        y_proba_doctor = doctor.predict_proba(X)
+        y_proba_net = doctor.net.predict_proba(X)
+        np.testing.assert_allclose(y_proba_doctor, y_proba_net)
+
+    def test_score(self, doctor, data):
+        X, y = data
+        score_doctor = doctor.score(X, y)
+        score_net = doctor.net.score(X, y)
+        np.testing.assert_allclose(score_doctor, score_net)
+
+    def test_recs_with_filter(self, module_cls, net_cls, doctor_cls, data, custom_split):
+        # when initializing SkorchDoctor with a match_fn, only records whose
+        # keys match should be kept
+        def match_fn(name):
+            return "lin0" in name
+
+        net = net_cls(module_cls, max_epochs=3, batch_size=32, train_split=custom_split)
+        doctor = doctor_cls(net, match_fn=match_fn)
+        doctor.fit(*data)
+
+        for rec in doctor.activation_recs_['module']:
```

Comment:
I know that an error would raise if nothing matched, but can we also assert `len(doctor.activation_recs_["module"])` to make sure there is something in the list?

Reply:
Done